### PR TITLE
Restore Boom friction code for demo compatibility

### PR DESCRIPTION
**Pull request by rfomin**
This fixes desync in cchest3.wad map05 [demos](https://dsdarchive.com/wads/cchest3?level=Map+05) and many others.

### CODE REVIEW

**Comment by rfomin** [comment]
After this change complevel 9 desyncs became rare, and all of them on extremely big slaughter maps. 
sunder.wad map13,  flotsam.wad map12, ngmvmt1.wad map13 for example. Is there a pattern here?

File: Source/p_map.c [hunk 1]
```diff
@@ -170,6 +170,48 @@ int P_GetMoveFactor(const mobj_t *mo, int *frictionp)
 {
   int movefactor, friction;
 
+  //e6y
```
Comment (fabiangreffrath) [inline Source/p_map.c:173]: This will require a more explicit comment than this.

File: Source/p_map.c [hunk 2]
```diff
@@ -1021,12 +1063,31 @@ static void P_HitSlideLine(line_t *ld)
   // Check for the special cases of horz or vert walls.
 
   // killough 10/98: only bounce if hit hard (prevents wobbling)
-  icyfloor = 
-    (demo_version >= 203 ? 
-     P_AproxDistance(tmxmove, tmymove) > 4*FRACUNIT : !compatibility) &&
-    variable_friction &&  // killough 8/28/98: calc friction on demand
-    slidemo->z <= slidemo->floorz &&
-    P_GetFriction(slidemo, NULL) > ORIG_FRICTION;
+  // icyfloor =
```
Comment (fabiangreffrath) [inline Source/p_map.c:1066]: Commented blocks with `/* ... */`, please.

File: Source/p_map.c [hunk 3]
```diff
@@ -1021,12 +1063,31 @@ static void P_HitSlideLine(line_t *ld)
...
+  //   (demo_version >= 203 ? 
+  //    P_AproxDistance(tmxmove, tmymove) > 4*FRACUNIT : !compatibility) &&
+  //   variable_friction &&  // killough 8/28/98: calc friction on demand
+  //   slidemo->z <= slidemo->floorz &&
+  //   P_GetFriction(slidemo, NULL) > ORIG_FRICTION;
+
+  //e6y
+  if (demo_version >= 203)
+  {
+    icyfloor =
```
Comment (fabiangreffrath) [inline Source/p_map.c:1076]: Most of this part of the code is actually identical to the original. If you keep indentation, there's probably only two conditions to change.

File: Source/p_mobj.c [hunk 4]
```diff
@@ -326,20 +333,41 @@ void P_XYMovement (mobj_t* mo)
...
+          }
+        mo->friction = ORIG_FRICTION; // reset to normal for next tic
+      }
+      else
+      {
+        fixed_t friction = P_GetFriction(mo, NULL);
 
-      mo->momx = FixedMul(mo->momx, friction);
-      mo->momy = FixedMul(mo->momy, friction);
+        mo->momx = FixedMul(mo->momx, friction);
```
Comment (fabiangreffrath) [inline Source/p_mobj.c:358]: If you keep the indentation original, the resulting code will look less clean, but the diff to the original code will be minimal. The latter is more important to me (especially when hunting demo sync bugs and comparing different states of code and where they actually differ).

File: Source/p_mobj.c [hunk 5]
```diff
@@ -769,6 +797,9 @@ mobj_t *P_SpawnMobj(fixed_t x, fixed_t y, fixed_t z, mobjtype_t type)
   mobj->thinker.function = P_MobjThinker;
   mobj->above_thing = mobj->below_thing = 0;           // phares
 
+  //e6y
```
Comment (fabiangreffrath) [inline Source/p_mobj.c:800]: Again, let's please not adapt bad practice of using three-letter comments that don't say anything about which code was changed and why.

File: Source/p_mobj.h [hunk 6]
```diff
@@ -336,6 +336,10 @@ typedef struct mobj_s
 
     // killough 8/2/98: friction properties part of sectors,
     // not objects -- removed friction properties from here
+    // e6y: restored friction properties here
```
Comment (fabiangreffrath) [inline Source/p_mobj.h:339]: Will this change break savegames?

File: Source/p_user.c [hunk 7]
```diff
@@ -102,14 +106,49 @@ void P_CalcHeight (player_t* player)
...
+  else
+  {
+    if (player->bob > MAXBOB)
+      player->bob = MAXBOB;
+  }
+
 
-  if (player->bob > MAXBOB)                             
-    player->bob = MAXBOB;
+  // if (player->bob > MAXBOB)
```
Comment (fabiangreffrath) [inline Source/p_user.c:150]: Please try to integrate the original code more cleanly into the additions

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]
Thanks for the effort, but these changes are so massive that we need to integrate them more cleanly into the existing code.

File: Source/p_mobj.h (hunk 6 above)
Comment (rfomin) [reply Source/p_mobj.h:339]: Yes, it breaks savegame. Not sure how to deal with it.

File: Source/p_mobj.c [hunk 8]
```diff
@@ -326,7 +332,27 @@ void P_XYMovement (mobj_t* mo)
...
+        else
+          { // Yes. Use stored friction.
+          mo->momx = FixedMul(mo->momx,mo->friction);
+          mo->momy = FixedMul(mo->momy,mo->friction);
+          }
+        mo->friction = ORIG_FRICTION; // reset to normal for next tic
+      }
+      else
+      {
+        fixed_t friction = P_GetFriction(mo, NULL);
```
Comment (fabiangreffrath) [inline Source/p_mobj.c:355]: Original indentation of the `fixed_t friction = P_GetFriction(mo, NULL);` line, please.

File: Source/p_user.c [hunk 9]
```diff
@@ -102,14 +105,24 @@ void P_CalcHeight (player_t* player)
...
+  if ((demo_version == 202 || 
+      demo_version == 203) &&
+      player->mo->friction > ORIG_FRICTION) // ice?
+  {
+    if (player->bob > (MAXBOB>>2))
+      player->bob = MAXBOB>>2;
+  }
+  else
+  {
+    if (player->bob > MAXBOB)
```
Comment (fabiangreffrath) [inline Source/p_user.c:123]: Original indentation for the `if (player->bob > MAXBOB)` block, please.

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]
Very much better now already. I prefer code changes formatted this way, so it's always obvious at first sight what was is different relative to the original code, even if this breaks with strict indentation rules.

File: Source/p_mobj.h (hunk 6 above)
Comment (fabiangreffrath) [reply Source/p_mobj.h:339]: This really nags me...

Comment (rfomin) [reply Source/p_mobj.h:339]: Yes, I think this is the last issue I tested about 5000 complevel 9 demos without desync.

**Pull request merged**